File: code/examples/Demo-app/android/DemoApp/src/main/kotlin/com/opencloudsheet/demo/viewmodel/SheetsViewModel.kt
package com.opencloudsheet.demo.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.opencloudsheet.OpenCloudSheetSdk
import com.opencloudsheet.Provider
import com.opencloudsheet.demo.model.Expense
import com.opencloudsheet.model.metadata.SheetMetadata
import com.opencloudsheet.model.workbook.IWorkBook
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for managing worksheets within a workbook.
 */
class SheetsViewModel(
    private val workbookId: String,
    private val workbookName: String
) : ViewModel() {

    private val _uiState = MutableStateFlow<SheetsUiState>(SheetsUiState.Loading)
    val uiState: StateFlow<SheetsUiState> = _uiState.asStateFlow()

    private var workbook: IWorkBook? = null

    init {
        loadWorkBook()
    }

    private fun loadWorkBook() {
        viewModelScope.launch {
            try {
                _uiState.value = SheetsUiState.Loading

                // Get workbook from metadata
                val workbooks = OpenCloudSheetSdk.getWorkBooks(Provider.OneDrive)
                val foundWorkbook = workbooks.find { it.getId() == workbookId }
                    ?: throw IllegalStateException("Workbook not found")

                workbook = foundWorkbook

                loadSheets()
            } catch (e: Exception) {
                _uiState.value = SheetsUiState.Error("Failed to load worksheet: ${e.message}")
            }
        }
    }

    private fun loadSheets() {
        viewModelScope.launch {
            try {
                if (workbook == null) {
                    _uiState.value = SheetsUiState.Error("Workbook not initialized")
                    return@launch
                }

                val sheets = workbook!!.getSheets()
                _uiState.value = SheetsUiState.Success(
                    workbookName = workbookName,
                    sheets = sheets
                )
            } catch (e: Exception) {
                _uiState.value = SheetsUiState.Error("Failed to load sheets: ${e.message}")
            }
        }
    }

    fun createSheet(month: String, year: Int) {
        viewModelScope.launch {
            try {
                _uiState.value = SheetsUiState.Creating

                if (workbook == null) {
                    _uiState.value = SheetsUiState.Error("Workbook not initialized")
                    return@launch
                }

                val sheetName = "$month $year"
                workbook!!.createSheet(
                    type = Expense::class.java,
                    name = sheetName,
                    description = "Expenses for $sheetName"
                )

                // Reload sheets
                loadSheets()
            } catch (e: Exception) {
                _uiState.value = SheetsUiState.Error("Failed to create sheet: ${e.message}")
            }
        }
    }

    fun deleteSheet(sheet: SheetMetadata) {
        viewModelScope.launch {
            try {
                if (workbook == null) {
                    _uiState.value = SheetsUiState.Error("Workbook not initialized")
                    return@launch
                }

                val currentState = _uiState.value
                if (currentState is SheetsUiState.Success) {
                    // Optimistically remove from UI
                    val updatedList = currentState.sheets.filter { it.sheetName != sheet.sheetName }
                    _uiState.value = SheetsUiState.Success(
                        workbookName = currentState.workbookName,
                        sheets = updatedList
                    )

                    // Delete from backend
                    workbook!!.deleteSheet(sheet.sheetName)
                }
            } catch (e: Exception) {
                _uiState.value = SheetsUiState.Error("Failed to delete sheet: ${e.message}")
                loadSheets()
            }
        }
    }

}

sealed class SheetsUiState {
    object Loading : SheetsUiState()
    object Creating : SheetsUiState()
    data class Success(
        val workbookName: String,
        val sheets: List<SheetMetadata>
    ) : SheetsUiState()
    data class Error(val message: String) : SheetsUiState()
}

File: code/examples/Demo-app/ios/CloudSheetDemo/CloudSheetDemo/ViewModels/SheetsViewModel.swift
//
//  SheetsViewModel.swift
//  CloudSheetDemo
//
//  ViewModel for managing worksheets within a workbook
//

import Foundation
import Combine
import OpenCloudSheet

enum SheetsUiState {
    case loading
    case creating
    case success(workbookName: String, sheets: [SheetMetadata])
    case error(message: String)
}

@MainActor
class SheetsViewModel: ObservableObject {
    @Published var uiState: SheetsUiState = .loading

    private let workbookId: String
    private let workbookName: String
    private var workbook: (any IWorkBook)?

    init(workbookId: String, workbookName: String) {
        self.workbookId = workbookId
        self.workbookName = workbookName
        loadWorkBook()
    }

    private func loadWorkBook() {
        uiState = .loading

        Task {
            do {
                let workbooks = try await OpenCloudSheetSdk.getWorkBooks(.OneDrive)
                guard let foundWorkbook = workbooks.first(where: { $0.getId() == workbookId }) else {
                    uiState = .error(message: "Workbook not found")
                    return
                }

                workbook = foundWorkbook
                loadSheets()
            } catch {
                uiState = .error(message: "Failed to load workbook: \(error.localizedDescription)")
            }
        }
    }

    private func loadSheets() {
        Task {
            do {
                guard let workbook = workbook else {
                    uiState = .error(message: "Workbook not initialized")
                    return
                }

                let sheets = try await workbook.getSheets()
                uiState = .success(workbookName: workbookName, sheets: sheets)
            } catch {
                uiState = .error(message: "Failed to load sheets: \(error.localizedDescription)")
            }
        }
    }

    func createSheet(month: String, year: Int) {
        uiState = .creating

        Task {
            do {
                guard let workbook = workbook else {
                    uiState = .error(message: "Workbook not initialized")
                    return
                }

                let sheetName = "\(month) \(year)"
                _ = try await workbook.createSheet(
                    type: Expense.self,
                    name: sheetName,
                    description: "Expenses for \(sheetName)"
                )
                loadSheets()
            } catch {
                uiState = .error(message: "Failed to create sheet: \(error.localizedDescription)")
            }
        }
    }

    func deleteSheet(sheet: SheetMetadata) {
        Task {
            do {
                guard let workbook = workbook else {
                    uiState = .error(message: "Workbook not initialized")
                    return
                }

                if case .success(let workbookName, let sheets) = uiState {
                    let updatedList = sheets.filter { $0.sheetName != sheet.sheetName }
                    uiState = .success(workbookName: workbookName, sheets: updatedList)

                    try await workbook.deleteSheet(name: sheet.sheetName)
                }
            } catch {
                uiState = .error(message: "Failed to delete sheet: \(error.localizedDescription)")
                loadSheets()
            }
        }
    }
}
